### PR TITLE
[XMLHttpRequest] Add support for ontimeout and onerror handler when using XMLHttpRequest for Android and iOS

### PR DESCRIPTION
**Pull request by grgmo**
Currently React-Native does not have `ontimeout` and `onerror` handlers for [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest). This is an extension to [No timeout on XMLHttpRequest](https://github.com/facebook/react-native/issues/4648).

With addition to two handlers, both Android and iOS can now handle `ontimeout` if request times out and `onerror` when there is general network error.

**Test plan**

Code has been tested on both Android and iOS with [Charles](https://www.charlesproxy.com/) by setting a breakpoint on the request which fires `ontimeout` when the request waits beyond `timeout` time and `onerror` when there is network error.

**Usage**

JavaScript - 

```
var request = new XMLHttpRequest();

function onLoad() {
    console.log(request.status);
};

function onTimeout() {
    console.log('Timeout');
};

function onError() {
    console.log('General network error');
};

request.onload = onLoad;
request.ontimeout = onTimeout;
request.onerror = onError;
request.open('GET', 'https://mywebsite.com/endpoint.php');
request.send();
```

### CODE REVIEW

**Comment by mkonicek** [comment]
I'll let @lexs review this one.

**Comment by mkonicek** [comment]
Do we have an XHR example in the UI Explorer? Can you add this there?

Can you add some unit / integration tests?

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 1]
```diff
@@ -326,6 +326,9 @@ private void onRequestError(ExecutorToken ExecutorToken, int requestId, String e
     WritableArray args = Arguments.createArray();
     args.pushInt(requestId);
     args.pushString(error);
+    if (error.equals("timeout")) {
```
Comment (lexs) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: Can you make this instead of using the string perhaps pass down the IOException? I'm not sure exactly what the exception type it'll be but perhaps `SocketTimeoutException`. Then just compare the `getClass()`. This avoids depending on an implementation detail somewhere setting "timeout" as the message.

It's probably easiest to split `onRequestError()` into two methods, one taking an error string and one taking an exception.

File: Libraries/Network/XMLHttpRequestBase.js [hunk 2]
```diff
@@ -296,6 +319,20 @@ class XMLHttpRequestBase {
       onload(null);
     }
   }
+
+  _sendOnTimeout(): void {
```
Comment (lexs) [inline Libraries/Network/XMLHttpRequestBase.js]: nit, you can consolidate these to something like `_sendEvent(this.ontimeout)` so you get rid of the duplication. then just call these from `setReadyState()`

**Comment by grgmo** [comment]
@lexs i've made the changes, added unit test and example on UIExplorer. Comments from eslint-bot are also fixed. I need to update the Network documentation if you're ok with the changes.

**Comment by lexs** [comment]
@facebook-github-bot shipit

**Comment by lexs** [comment]
Thanks! Looks great, sorry for taking a few days coming back to this.

**Comment by lexs** [comment]
@grgmo Seems like there's merge issue, could you rebase and I'll reimport?

**Comment by grgmo** [comment]
@lexs on it, will update the doc too. thanks for reviewing

**Comment by skellock** [comment]
Thank you patchers and mergers.  <3

**Comment by skellock** [comment]
@grgmo i know this PR is wrapping up, and the following question changes nothing, but i'm curious:  does onerror bubble up some semblance of a error code?   

I ask because in my testing, i've seen the error messages appear differently on different devices:

For example:

```
<= iOS 8.2 The operation couldn’t be completed. (NSURLErrorDomain error -1001.)
>= iOS 8.3 The request timed out.
android 22 - SSL handshake timed out (if a low timeout is set and server is ssl)
android 22 - timeout (if a high timeout is set)
android 16 - Read timed out (if a high timeout is set)
```

Not just timeouts too.  Also non-http server errors.  Seems like iOS 8.2 and under at least manifested the underlying error code, but that stopped at 8.3.

And Android... well... ¯\_(ツ)_/¯

TL;DR:  Any chance we could harvest the underlying error codes?  I wouldn't mind tackling that work in native side, but it feels a bit weird to decorate XMLHttpRequest with extra things that aren't in the spec.  (potentially over/underthinking it).

**Comment by lexs** [comment]
@skellock Yeah, problem with network errors is that there is no real standard way of expressing them as with HTTP error codes (404 etc). We could try to consolidate them (by exception type?) but as you said the XMLHttpRequest API isn't very granular and apart from `ontimeout` and the like.

**Comment by skellock** [comment]
Well this PR is a beautiful first step.  Now that I know the difference between `timeout` and `other`, that's good enough to relay better info on to the user.  Thank you.

**Comment by grgmo** [comment]
@lexs rebased

**Comment by lexs** [comment]
@facebook-github-bot shipit
